File: library/src/test/java/com/wultra/android/sslpinning/CertStoreValidationTest.java
/*
 * Copyright 2018 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 */

package com.wultra.android.sslpinning;

import com.wultra.android.sslpinning.model.GetFingerprintResponse;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.modules.junit4.PowerMockRunner;

import java.net.URL;
import java.security.cert.X509Certificate;
import java.util.Base64;
import java.util.Date;
import java.util.concurrent.TimeUnit;

import static org.junit.Assert.assertEquals;

/**
 * Tests for validation with [CertStore].
 *
 * @author Tomas Kypta, tomas.kypta@wultra.com
 */
@RunWith(PowerMockRunner.class)
public class CertStoreValidationTest extends CommonJavaTest {

    @Test
    public void testValidationGithubFallbackValid() throws Exception {
        // necessary to update the fingerprint from time to time
        // Mar 2023 - "kqN/vV4hpTqVxxbhFE9EL1grlND6/Gc+tnF6TrUaiKc="
        validateGithubWithFallbackOnly("kqN/vV4hpTqVxxbhFE9EL1grlND6/Gc+tnF6TrUaiKc=", ValidationResult.TRUSTED);
    }

    @Test
    public void testValidationGithubFallbackInvalid() throws Exception {
        // outdated fingerprint
        validateGithubWithFallbackOnly("trmmrz6GbL4OajB+fdoXOzcrLTrD8GrxX5dxh3OEgAg=", ValidationResult.UNTRUSTED);
    }

    private void validateGithubWithFallbackOnly(String fingerprintBase64,
                                                ValidationResult expectedResult) throws Exception {
        X509Certificate cert = TestUtils.getCertificateFromUrl("https://github.com");

        String publicKey = "BC3kV9OIDnMuVoCdDR9nEA/JidJLTTDLuSA2TSZsGgODSshfbZg31MS90WC/HdbU/A5WL5GmyDkE/iks6INv+XE=";
        byte[] publicKeyBytes = Base64.getDecoder().decode(publicKey);

        String signatureBase64 = "MEUCICB69UpMPOdtrsR6XcJqHEh2L2RO4oSJ3SZ7BYnTBJbGAiEAnZ7rEWdMVGwa59Wx5QbAorEFxXH89Iu0CnqWa96Eda0=";
        byte[] signatureBytes = Base64.getDecoder().decode(signatureBase64);
        byte[] fingerprintBytes = Base64.getDecoder().decode(fingerprintBase64);

        GetFingerprintResponse.Entry fallbackEntry = new GetFingerprintResponse.Entry(
                "github.com",
                fingerprintBytes,
                new Date(System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1)),
                signatureBytes);
        GetFingerprintResponse fallback = new GetFingerprintResponse(new GetFingerprintResponse.Entry[] {fallbackEntry});

        CertStoreConfiguration config = TestUtils.getCertStoreConfiguration(
                new Date(),
                new String[]{"github.com"},
                new URL("https://github.com"),
                publicKeyBytes,
                fallback);
        CertStore store = new CertStore(config, cryptoProvider, secureDataStore);
        TestUtils.assignHandler(store, handler);
        ValidationResult result = store.validateCertificate(cert);
        assertEquals(expectedResult, result);
    }

    @Test
    public void testValidationGithubUpdateWithOutdatedData() throws Exception {
        // json with outdated data, correct public key
        validateGithubWithUpdateJsonOnly("BC3kV9OIDnMuVoCdDR9nEA/JidJLTTDLuSA2TSZsGgODSshfbZg31MS90WC/HdbU/A5WL5GmyDkE/iks6INv+XE=",
                "https://gist.githubusercontent.com/hvge/7c5a3f9ac50332a52aa974d90ea2408c/raw/34866234bbaa3350dc0ddc5680a65a6f4e7c549e/ssl-pinning-signatures.json",
                UpdateResult.STORE_IS_EMPTY, ValidationResult.EMPTY);
    }
    @Test
    public void testValidationGithubUpdateWithInvalidSignature() throws Exception {
        // json with current data, different public key
        validateGithubWithUpdateJsonOnly("BC3kV9OIDnMuVoCdDR9nEA/JidJLTTDLuSA2TSZsGgODSshfbZg31MS90WC/HdbU/A5WL5GmyDkE/iks6INv+XE=",
                "https://gist.githubusercontent.com/TomasKypta/40be50cc63d2f4c00abcbbf4554f0e32/raw/9cc9029d9e8248b0cd9a36b98382040114dd1d4a/ssl-pinning-signatures_Mar2023.json",
                UpdateResult.INVALID_SIGNATURE, ValidationResult.EMPTY);
    }

    @Test
    public void testValidationGithubUpdateWithValidData() throws Exception {
        // json with current data, correct public key
        validateGithubWithUpdateJsonOnly("BC3kV9OIDnMuVoCdDR9nEA/JidJLTTDLuSA2TSZsGgODSshfbZg31MS90WC/HdbU/A5WL5GmyDkE/iks6INv+XE=",
                "https://gist.githubusercontent.com/hvge/7c5a3f9ac50332a52aa974d90ea2408c/raw/07eb5b4b67e63d37d224912bc5951c7b589b35e6/ssl-pinning-signatures.json",
                UpdateResult.OK, ValidationResult.TRUSTED);
    }

    private void validateGithubWithUpdateJsonOnly(String publicKey,
                                                  String signaturesJsonUrl,
                                                  UpdateResult expectedUpdateResult,
                                                  ValidationResult expectedValidationResult) throws Exception {
        X509Certificate cert = TestUtils.getCertificateFromUrl("https://github.com");

        byte[] publicKeyBytes = Base64.getDecoder().decode(publicKey);

        CertStoreConfiguration config = TestUtils.getCertStoreConfiguration(
                new Date(),
                new String[]{"github.com"},
                new URL(signaturesJsonUrl),
                publicKeyBytes,
                null);
        CertStore store = new CertStore(config, cryptoProvider, secureDataStore);
        TestUtils.assignHandler(store, handler);
        TestUtils.updateAndCheck(store, UpdateMode.FORCED, expectedUpdateResult);

        ValidationResult result = store.validateCertificate(cert);
        assertEquals(expectedValidationResult, result);
    }
}

File: library/src/test/java/com/wultra/android/sslpinning/CertStoreValidationTest.kt
/*
 * Copyright 2018 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 */

package com.wultra.android.sslpinning

import com.wultra.android.sslpinning.model.GetFingerprintResponse
import com.wultra.android.sslpinning.service.RemoteDataProvider
import com.wultra.android.sslpinning.service.RemoteDataResponse
import io.mockk.every
import io.mockk.mockk

import org.junit.Test

import java.net.URL
import java.util.Base64
import java.util.Date
import java.util.concurrent.TimeUnit

import org.junit.Assert.assertEquals

/**
 * Tests for validation with [CertStore].
 *
 * @author Tomas Kypta, tomas.kypta@wultra.com
 */
class CertStoreValidationTest : CommonKotlinTest() {

    @Test
    @Throws(Exception::class)
    fun testValidationGithubFallbackInvalid() {
        val fingerprintBase64 = "trmmrz6GbL4OajB+fdoXOzcrLTrD8GrxX5dxh3OEgAg="
        val expectedResult = ValidationResult.UNTRUSTED
        val publicKey = "BC3kV9OIDnMuVoCdDR9nEA/JidJLTTDLuSA2TSZsGgODSshfbZg31MS90WC/HdbU/A5WL5GmyDkE/iks6INv+XE="
        val publicKeyBytes = Base64.getDecoder().decode(publicKey)

        val signatureBase64 = "MEUCICB69UpMPOdtrsR6XcJqHEh2L2RO4oSJ3SZ7BYnTBJbGAiEAnZ7rEWdMVGwa59Wx5QbAorEFxXH89Iu0CnqWa96Eda0="
        val signatureBytes = Base64.getDecoder().decode(signatureBase64)
        val fingerprintBytes = Base64.getDecoder().decode(fingerprintBase64)

        val fallbackEntry = GetFingerprintResponse.Entry(
            "github.com",
            fingerprintBytes,
            Date(System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1)),
            signatureBytes)
        val fallback = arrayOf(fallbackEntry)

        val config = TestUtils.getCertStoreConfiguration(
            Date(),
            arrayOf("github.com"),
            URL("https://github.com"),
            publicKeyBytes,
            fallback)
        val store = CertStore(config, cryptoProvider, secureDataStore)
        TestUtils.assignHandler(store, handler)
        val result = store.validateCertificate(TestUtils.githubCert())
        assertEquals(expectedResult, result)
    }

    @Test
    @Throws(Exception::class)
    fun testValidationGithubUpdateWithOutdatedData() {

        val remoteDataProvider: RemoteDataProvider = mockk()
        every { remoteDataProvider.getFingerprints(any()) } answers {
            RemoteDataResponse(
                200,
                sigHeader,
                """
                    {
                      "fingerprints": [
                        {
                          "name" : "github.com",
                          "fingerprint" : "trmmrz6GbL4OajB+fdoXOzcrLTrD8GrxX5dxh3OEgAg=",
                          "expires" : 1652184000,
                          "signature" : "MEUCIQCs1y/nyrKh4+2DIuX/PufUYiaVUdt2FBZQg6rBeZ/r4QIgNlT4owBwJ1ThrDsE0SwGipTNI74vP1vNyLNEwuXY4lE="
                        }
                      ]
                    }
                """.toByteArray()
            )
        }
        every { cryptoProvider.ecdsaValidateSignature(any(), any()) } returns true

        // json with outdated data, correct public key
        validateGithubWithUpdateJsonOnly(remoteDataProvider, UpdateResult.OK, ValidationResult.EMPTY)
    }

    @Test
    @Throws(Exception::class)
    fun testValidationGithubUpdateWithInvalidSignature() {

        val remoteDataProvider: RemoteDataProvider = mockk()
        every { remoteDataProvider.getFingerprints(any()) } answers {
            RemoteDataResponse(
                200,
                emptyMap(),
                """
                    { 
                  "fingerprints": [{
                    "name" : "github.com",
                    "fingerprint" : "kqN/vV4hpTqVxxbhFE9EL1grlND6/Gc+tnF6TrUaiKc=",
                    "expires" : 2212460799,
                    "signature" : "MEQCIAiq/O5IbZ8K2SsZtDbpsvHWecxu4eLOSS7oOXjDk2KeAiBKkI53ESVByK+wKwaLA5LsEu8oonUHiYVM2zQtWf46DA=="
                  }]
                }
                """.toByteArray()
            )
        }
        every { cryptoProvider.ecdsaValidateSignature(any(), any()) } returns true

        // json with current data, invalid header signature
        validateGithubWithUpdateJsonOnly(remoteDataProvider, UpdateResult.INVALID_SIGNATURE, ValidationResult.EMPTY)
    }

    @Test
    @Throws(Exception::class)
    fun testValidationGithubUpdateWithValidData() {

        val remoteDataProvider: RemoteDataProvider = mockk()
        every { remoteDataProvider.getFingerprints(any()) } answers {
            RemoteDataResponse(
                200,
                sigHeader,
                TestUtils.validFingerprintJsonResponse()
            )
        }
        every { cryptoProvider.ecdsaValidateSignature(any(), any()) } returns true
        
        // json with current data, correct public key
        validateGithubWithUpdateJsonOnly(remoteDataProvider, UpdateResult.OK, ValidationResult.TRUSTED)
    }

    @Throws(Exception::class)
    fun validateGithubWithUpdateJsonOnly(
        provider: RemoteDataProvider,
        expectedUpdateResult: UpdateResult,
        expectedValidationResult: ValidationResult
    ) {
        val publicKeyBytes = Base64.getDecoder().decode(TestUtils.testPubKey())
        val config = TestUtils.getCertStoreConfiguration(
            Date(),
            arrayOf("github.com"),
            URL("https://test"),
            publicKeyBytes,
            null
        )

        val store = CertStore(config, cryptoProvider, secureDataStore, provider)

        TestUtils.assignHandler(store, handler)
        TestUtils.updateAndCheck(store, UpdateMode.FORCED, expectedUpdateResult)

        val result = store.validateCertificate(TestUtils.githubCert())
        assertEquals(expectedValidationResult, result)
    }
}
